//protoc -I . --python_out=. --grpc_out=.      --plugin=protoc-gen-grpc=/grpc/bins/opt/grpc_python_plugin playcloud.proto
//protoc -I . --java_out=.   --grpc-java_out=. --plugin=protoc-gen-grpc-java=grpc-java/compiler/build/binaries/java_pluginExecutable/protoc-gen-grpc-java playcloud.proto
syntax = "proto3";
option java_package = "ch.unine.iiun.safecloud";

//the encoder service
service EncoderDecoder {
	rpc Encode (EncodeRequest) returns (EncodeReply) {}
	rpc Decode (DecodeRequest) returns (DecodeReply) {}
	rpc Reconstruct (ReconstructRequest) returns (ReconstructReply) {}
}

//the proxy service
service Proxy {
	// Returns at most n random blocks where n is a parameter of the function
	rpc GetRandomBlocks (BlockRequest) returns (BlockReply) {};
	rpc GetBlock (NamedBlockRequest) returns (Strip) {};
}

//Data types
message Strip {
	bytes data = 1;
	bytes checksum = 2;
}
message EncodeRequest {
	bytes payload = 1;
	map<string,string> parameters = 2;
}

message EncodeReply {
	repeated Strip strips = 1;
	map<string,string> parameters = 2;
}

message DecodeRequest {
	repeated Strip strips = 1;
	map<string, string> parameters = 2;
}

message DecodeReply {
	bytes dec_block = 1;
	map<string,string> parameters = 2;
}

message BlockRequest {
  uint32 blocks = 1;
}

message BlockReply {
  repeated Strip strips = 1;
}

message NamedBlockRequest {
	string path = 1;
	uint32 index = 2;
}

message ReconstructRequest {
	string path = 1;
	repeated uint32 missing_indices = 2;
}

message ReconstructReply {
	map<uint32, Strip> reconstructed = 1;
}
